cmd/activator/round_trippers_test.go

```diff
@@ -0,0 +1,152 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+func TestRetryRoundTripper(t *testing.T) {
+	wantBody := "all good!"
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(wantBody))
+	}))
+	l := zap.NewExample().Sugar()
+	maxRetries := 3
+	interval := 10 * time.Millisecond
+
+	examples := []struct {
+		label   string
+		retries int
+		wantErr bool
+	}{
+		{"success", maxRetries, false},
+		{"failure", maxRetries + 1, true},
+	}
+
+	for _, e := range examples {
+		t.Run(e.label, func(t *testing.T) {
+			var last time.Time
+
+			gotRetries := 0
+			rt := roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+				gotRetries += 1
+
+				now := time.Now()
+				duration := now.Sub(last)
+				if duration < interval {
+					t.Errorf("Unexpected retry interval. Want %v, got %v", interval, duration)
+				}
+				last = now
+
+				if gotRetries < e.retries {
+
+					if r.Body != nil {
+						ioutil.ReadAll(r.Body)
+						r.Body.Close()
+					}
+
+					return nil, errors.New("some error!")
+				}
+
+				return http.DefaultTransport.RoundTrip(r)
+			})
+
+			rrt := newRetryRoundTripper(rt, l, maxRetries, interval)
+			req := httptest.NewRequest("", ts.URL, nil)
+
+			resp, err := rrt.RoundTrip(req)
+
+			wantRetries := maxRetries
+			if e.retries < wantRetries {
+				wantRetries = e.retries
+			}
+
+			if gotRetries != wantRetries {
+				t.Errorf("Unexpected number of retries. Want %d, got %d", wantRetries, gotRetries)
+			}
+
+			if e.wantErr {
+				if err == nil {
+					t.Errorf("Expected error")
+				}
+			} else {
+				if err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+				gotBody, _ := ioutil.ReadAll(resp.Body)
+				if string(gotBody) != wantBody {
+					t.Errorf("Unexpected response. Want %q, got %q", wantBody, gotBody)
+				}
+			}
+		})
+	}
+}
+
+func TestStatusFilterRoundTripper(t *testing.T) {
+	testServer := func(status int) *httptest.Server {
+		return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(status)
+		}))
+	}
+
+	goodRT := http.DefaultTransport
+	errorRT := roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+		return nil, errors.New("some error")
+	})
+
+	filtered := []int{501, 502}
+
+	examples := []struct {
+		label     string
+		transport http.RoundTripper
+		status    int
+		err       error
+	}{
+		{"filtered status", goodRT, 502, errors.New("Filtering 502")},
```

Comment:
Test cases here seem to be too limited. Can we add tests to cover all the cases listed below?

https://github.com/mdemirhan/serving/blob/2bf10f2fca8e56ea174b3259287f62bc2385f26c/pkg/activator/httptransport_test.go#L81

Reply:
Sure, I'll add test cases for http1/http2. I think everything else is covered in other, separate tests?

Reply:
Actually there are a few more. List of test cases to add so far:

* http1/http2
* nil response body

Reply:
Is the nil response body check actually necessary? From the source, Response Body is guaranteed to be non-nil: 

https://github.com/golang/go/blob/master/src/net/http/response.go#L57-L62